File: frontend/src/lib/otel.ts
/**
 * OpenTelemetry initialization for browser tracing.
 *
 * Runtime-gated: only activates when window.__OTEL_ENDPOINT__ is set
 * (injected by nginx/lua at serve time). Always bundled, no-op when
 * the endpoint is absent.
 */

import { WebTracerProvider } from '@opentelemetry/sdk-trace-web';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { resourceFromAttributes } from '@opentelemetry/resources';
import { trace } from '@opentelemetry/api';
import { ZoneContextManager } from '@opentelemetry/context-zone';

declare global {
  interface Window {
    __OTEL_ENDPOINT__?: string;
  }
}

let initialized = false;

/**
 * Initialize OpenTelemetry tracing if the collector endpoint is configured.
 * Safe to call multiple times -- subsequent calls are no-ops.
 */
export function initOtel(): void {
  if (initialized) return;

  const endpoint = window.__OTEL_ENDPOINT__;
  if (!endpoint) return;

  const resource = resourceFromAttributes({
    'service.name': 'ai-resume-frontend',
  });

  const provider = new WebTracerProvider({ resource });

  const exporter = new OTLPTraceExporter({
    url: `${endpoint}/v1/traces`,
  });

  provider.addSpanProcessor(new SimpleSpanProcessor(exporter));
  provider.register({
    contextManager: new ZoneContextManager(),
  });

  initialized = true;
}

/**
 * Return a tracer instance for creating spans.
 */
export function getTracer(name = 'ai-resume-frontend') {
  return trace.getTracer(name);
}

/**
 * Build a W3C traceparent header value from the currently active span.
 * Returns null when tracing is inactive or no span is in context.
 */
export function getTraceparent(): string | null {
  const span = trace.getActiveSpan();
  if (!span) return null;

  const ctx = span.spanContext();
  if (!ctx.traceId || ctx.traceId === '00000000000000000000000000000000') {
    return null;
  }

  const flags = ctx.traceFlags.toString(16).padStart(2, '0');
  return `00-${ctx.traceId}-${ctx.spanId}-${flags}`;
}

/**
 * Reset internal state. Exported only for testing.
 */
export function _resetForTesting(): void {
  initialized = false;
}

File: api-service/ai_resume_api/otel.py
"""OpenTelemetry initialization, gated on OTEL_EXPORTER_OTLP_ENDPOINT.

When the environment variable OTEL_EXPORTER_OTLP_ENDPOINT is set, this module
configures a TracerProvider with OTLP gRPC export and instruments the FastAPI
application. When the variable is absent, all calls are no-ops -- the default
OpenTelemetry NoOp tracer is used and zero overhead is added to request paths.
"""

import os
from typing import Any

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

_otel_initialized = False


def init_otel(app: Any = None) -> None:
    """Initialize OpenTelemetry if OTEL_EXPORTER_OTLP_ENDPOINT is set.

    Safe to call multiple times -- subsequent calls are no-ops.

    Args:
        app: FastAPI application instance to auto-instrument. If None,
             only the global TracerProvider is configured.
    """
    global _otel_initialized

    endpoint = os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT")
    if not endpoint:
        return

    if _otel_initialized:
        if app:
            FastAPIInstrumentor.instrument_app(app)
        return

    resource = Resource.create(
        {
            "service.name": os.environ.get("OTEL_SERVICE_NAME", "ai-resume-api"),
        }
    )

    provider = TracerProvider(resource=resource)
    exporter = OTLPSpanExporter(endpoint=endpoint, insecure=True)
    provider.add_span_processor(BatchSpanProcessor(exporter))
    trace.set_tracer_provider(provider)

    if app:
        FastAPIInstrumentor.instrument_app(app)

    _otel_initialized = True


def get_tracer(name: str = "ai-resume-api") -> trace.Tracer:
    """Get a tracer instance.

    Returns the global tracer which is either a real SDK tracer (when OTel is
    initialized) or the default NoOp tracer.
    """
    return trace.get_tracer(name)
